### ISSUE TITLE
Implement namespaces

### ISSUE THREAD

**Issue by me2beats** [godotengine/godot-proposals#1566]
**Describe the project you are working on:**
GDScript plugins

**Describe the problem or limitation you are having in your project:**
`class_name` is a great feature and I use it often when making plugins.

But there are at least 3 problems which can occur when using `class_name`:
1. If you use a custom class named say `Util` when creating a plugin, there is absolutely no guarantee that the user is not using a custom type with the same name.
So you will get
    > Unique global class "Utils" already exists

2. As the number of your custom classes gets large (imagine a room with several hundred items on the floor), you will most likely want to structure them by dividing them into logical parts (arrange all items on their shelves) for easier access. But in this case, you will face various challenges, and it seems that none of the solutions will be truly convenient
especially if you want to keep having each class separately (i.e. don't use nested classes)

3. Littering the global namespace as a whole.
all custom classes are visible in the global namespace. The user might mistakenly try to use your class.
also a large number of custom classes clogs up the autocomplete.

thus, having a large number of custom classes is incompatible with ease of development, which, as the code base of the project grows, will force the developer to abandon their use or resort to hackу workarounds.

However, this deprives the user of the benefits of using `class_name`.
Perhaps custom classes are the most convenient way to have classes as separate files that need global access.
It's just convenient. For example, I have many utility classes with static methods as custom classes.

The problem is especially acute when creating plugins.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
Implementing namespaces system could solve the ploblem.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
I imagine it as a new keyword - `namespace`
For example
`a.gd:`
```gdscript
namespace X
class_name A
```

`b.gd:`
```gdscript
namespace X
class_name B
```

in this case, access to classes A and B could be done as follows:
`new_script.gd`:
```gdscript
func _ready():
  print (X.A)
  print (X.B)
```

when trying to get class `A` from the global namespace, an error will appear:
`new_script.gd`:
```gdscript
func _ready():
  print (A)
```
> Error: The identifier "A" isn't declared in the current scope

As you can see, several scripts can use one namespace, which is one of the main advantages of namespaces over custom classes

using the same class inside the same namespace should lead to error:
`a.gd:`
```gdscript
namespace X
class_name A
```

`a_new.gd:`
```gdscript
namespace X
class_name A
```
> Unique global class "A" already exists in namespace: X

Another important advantage of namespaces is the ability to use nested namespaces.

`c.gd:`
```gdscript
namespace X.Y
class_name C
```
`new_script.gd`:
```gdscript
func _ready():
  print (X.Y.C)
```

If namespace is not specified (keyword `namespace` is not used), then just the global namespace is used (as usual).
That is, when creating a script `d.gd` with `class_name D`, class D will be visible in any script.
`d.gd:`
```gdscript
# namespace not specified, use the usual scenario
class_name D
```
`new_script.gd`:
```gdscript
func _ready():
  print (D)
```

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
No it can't

**Is there a reason why this should be core and not an add-on in the asset library?:**
Implementing this feature not at the core level would hardly be a good solution.

**Comment by Xrayez** [comment]
> **If this enhancement will not be used often, can it be worked around with a few lines of script?:**
> No it can't

> especially if you want to keep having each class separately (i.e. don't use nested classes)

On the project level, it's kind of possible, since you can create a single global class to contain other classes (not necessarily inner classes), see concrete example in [`WAT/namespace.gd`](https://github.com/CodeDarigan/WAT/blob/fa509c093dbbd1bb71f2f075aa091aa6bdd8667e/addons/WAT/namespace.gd).

```gdscript
# utils.gd
extends Node

static func say_hello():
	print("Hello!")
```

```gdscript
# namespace.gd
class_name Game extends Reference

const Utils = preload("res://utils.gd")

```

```gdscript
# main.gd
extends Node2D

func _ready():
	Game.Utils.say_hello()
```

Not sure how efficient such solution would be compared to just using `class_name`, since it takes an extra "namespace resolution" with such approach.

But yeah namespace issues in general are must have to resolve if we aim for proper plugins support, and we have similar issues with C++ namespaces when it comes to exposing core classes to scripting, with some solutions: godotengine/godot#26990.

**Comment by Calinou** [comment]
As a workaround, you can add a prefix to your class names if you expect your class names to conflict with other plugins (or user code).

That said, if the class name isn't a single proper noun, it makes conflicts much less likely already.

**Comment by CodeDarigan** [comment]
> 
> > **If this enhancement will not be used often, can it be worked around with a few lines of script?:**
> > No it can't
> 
> > especially if you want to keep having each class separately (i.e. don't use nested classes)
> 
> On the project level, it's kind of possible, since you can create a single global class to load other classes (not necessarily inner classes), see concrete example in [`WAT/namespace.gd`](https://github.com/CodeDarigan/WAT/blob/fa509c093dbbd1bb71f2f075aa091aa6bdd8667e/addons/WAT/namespace.gd).


@Xrayez To an extent you can do that but 1) Cyclic References, so you can't be comprehensive and 2) This can still conflict with inner classes of the classes in that namespace, such as this [state class conflict](https://github.com/CodeDarigan/WAT/issues/150) issue.

Namespaces are essentially mandatory for plugin creators.

**Comment by Xrayez** [comment]
Namespace collisions with already mentioned testing frameworks: bitwes/Gut#215. 😃

**Comment by Spartan322** [comment]
Also makes it more comparable to C# and increases organization of projects without cluttering the global space, tho at this point wouldn't it be better to stop relying upon files as classes? Its already a chore that if you need to keep using a bunch of classes, every single one has to have a class_name or preload (which tbh ends up requiring load most of the time because cyclical dependency issues) so with any moderately sized project you not only start running into global bloat in code that becomes difficult to manage, (especially without a dedicated IDE for GDScript) but massive amount of boilerplate that seems counterproductive.

**Comment by Xrayez** [comment]
See previous discussion in godotengine/godot#21215.

**Comment by FrederickDesimpel** [comment]
Really would like to see this happening.

I'm staying away from class_name (name_global ?) because of this, and also find the cyclic dependencys are a headache.

Why can't I type a member in a class to itself ( say to have a ref to a parent of the same type ), if only for typing/autocomplete. As long as you don't instance it i don't see why that would give a cyclic dependency. 

I tried all kinds of things... autoload with nodes seems to work well, but it always breaks at the last step. haven't tried one global class_name as the namespace yet. I don't think you can do nested namespace (name_space ?) with that workaround either ?

Tried a const preloaded .tres too, since that is loaded only once...  but cyclic again

> As a workaround, you can add a prefix to your class names
A workaround indeed. With proper namespace support you can have some prefix like author or studio initials or whatever as the first part.

**Comment by FrederickDesimpel** [comment]
Also, when considering name_spaces, it would be nice to be able to have 'imports' with shorthands, like in

import My_Stuff.Nested.NameSpace.Class as MyClass

SomeObject : Myclass
Myclass.whatever()

which would basically be like working with the consts now.

**Comment by PranavSK** [comment]
We could have a namespace approach if we can implement this #1709 where the folder structure could be determined at compile time and use that as namespaces. This also works well with Godot’s approach of organising assets by scene and logic.

**Comment by dueddel** [comment]
I appreciate the idea of introducing namespaces. I come from other languages having namespaces and already wondered if this is possible in GDScript. Found out it isn't. That being said, I upvote your request, @me2beats .

For now I go for the workaround of prefixing my classnames as it has been done 10 years ago in PHP, for instance.

So, I'd have something like `class_name MyDomain_Util_WhateverThisClassIsCalled`. Not very beautiful, but at least avoiding name collisions.

The workaround as suggested by @Xrayez is somehow neat and odd at the same time. "Neat" because I like how the classes can be used, like `MyDomain.Util.WhateverThisClassIsCalled.new()` which looks by far better than using underscores. But "odd" because of writing the extra classes to achieve these "namespaces" is kinda intricate.

As of writing this I am not sure what feels better in the end. 🤔

Just make namespaces happen! 😉 

Also, I agree to @FrederickDesimpel. If namespaces are introduced to GDScript, some `import` feature is needed as well.

**Comment by quantumcode-martin** [comment]
Anything new about namespaces ?

**Comment by Calinou** [comment]
> Anything new about namespaces ?

GDScript in 4.0 is in feature freeze as the goal is now to stabilize it. Therefore, no new features will be added to it until 4.1 at least.

**Comment by me2beats** [comment]
Changed to to `Implement namespaces` (removed `to avoid collisions in third-party add-ons`)
because add-ons is not the only reason to have the feature.

For example you have many classes and you want to group them.
Refactorings scripts so that they will use classes as namespaces may be a so-so idea, because you may not want to have big scripts

**Comment by red1939** [comment]
@AlexDarigan Does is mean casting to such `const X = preload()` type will work now?

**Comment by AlexDarigan** [comment]
> @AlexDarigan Does is mean casting to such `const X = preload()` type will work now?

Seems like it.

**Comment by Pyxus** [comment]
To add to the discussion. Using class names and preloads to fake namespaces is the approach I've taken in my addon, and it works well enough. However, in 4.0 class names seem to be required to take advantage of the new custom resource exports, and the doc comment's class referencing. I suppose that would be resolved if they were changed to not require class names but namespaces in GDScript would still be ideal.

**Comment by AlexDarigan** [comment]
Yeah. I've come across the issue with documentation comments. I don't think it is an unfair ask for plugin developers have some way to prevent polluting the global namespace.

**Comment by derkork** [comment]
You also need to have a `class_name` if your want your custom nodes to be showing in the node tree in the "Add node" dialog.

**Comment by derkork** [comment]
> I suppose that would be resolved if they were changed to not require class names but namespaces in GDScript would still be ideal.

It would still be a problem if user's of the addon would need to use the class name in their scripts. E.g. if you have a class `Flubbermator` in your addon which is a node and the users would want to use this in their script like this:

```swift
@onready var flubbermator:Flubbermator = %Flubbermator
```

then they need a proper class name unless they also do this `preload` thing which I personally find hard to maintain. When using class_name it doesn't matter where my file resides, but using `preload` or `load` requires me to hardcode the path which makes every move of a script a breaking change. Ultimately I think namespaces are the only real solution to this. A workaround is prefixing all addon classes with something which doesn't help readability.

**Comment by nice-shot** [comment]
Adding to what @Pyxus wrote, it's very possible to effectively have namespaces by using the `@preload` keyword but the fact that custom resource exports isn't available means that it's extremely limited for creating things like component systems.

A simple use case for me will be:

```
const HealthComponent = preload("./health/health_component.gd")

@export var health: HealthComponent
```

However, in the editor this will not let me select the relevant node. Only by using `class_name` will the node be selectable.

![ExportCustomComponent](https://github.com/godotengine/godot-proposals/assets/2602328/24c958c6-bc46-460f-9e4f-000fc635db1e)

Is it possible to allow this method to work without adding the more complex namespaces to gdscript?

**Comment by adamscott** [comment]
> Adding to what @Pyxus wrote, it's very possible to effectively have namespaces by using the `@preload` keyword but the fact that custom resource exports isn't available means that it's extremely limited for creating things like component systems.
> 
> A simple use case for me will be:
> 
> ```
> const HealthComponent = preload("./health/health_component.gd")
> 
> @export var health: HealthComponent
> ```
> 
> However, in the editor this will not let me select the relevant node. Only by using `class_name` will the node be selectable.
> 
> ![ExportCustomComponent](https://private-user-images.githubusercontent.com/2602328/272418001-24c958c6-bc46-460f-9e4f-000fc635db1e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY3MTU0NjYsIm5iZiI6MTcwNjcxNTE2NiwicGF0aCI6Ii8yNjAyMzI4LzI3MjQxODAwMS0yNGM5NThjNi1iYzQ2LTQ2MGYtOWU0Zi0wMDBmYzYzNWRiMWUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEzMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMzFUMTUzMjQ2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTAwMTcxZmQ4MzNjMzA1ZmQzZGI2MTlhMWIwNTRiZTJhOWE3MDMwZTMxYWUzYTY5YTc1NmRmZTYzOGI3Yjc4ZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.wwQcQtGVJUfXdjcwSHUvy7Kwy5fNxoYl9XBARhp9Nx8)
> 
> Is it possible to allow this method to work without adding the more complex namespace concept to gdscript?

Could you try again? I tried in 4.2 and it works flawlessly.

![image](https://github.com/godotengine/godot-proposals/assets/270928/869e1e5f-e351-4947-9aa9-a88b28d36cd4)

MRP: [home-namespaces.zip](https://github.com/godotengine/godot-proposals/files/14113819/home-namespaces.zip)

**Comment by nice-shot** [comment]
> Could you try again? I tried in 4.2 and it works flawlessly.

This does not work. I looked at your MRP and you've added `class_name` to health_component.gd so it only works because of that.

**Comment by hayahane** [comment]
It seems like `@export` doesn't allow non-global class to be exported in 4.3beta2. This makes namespace a needed feature.
![image](https://github.com/godotengine/godot-proposals/assets/39263675/44b54baf-5c81-4ae5-93a7-19e078dbd2eb)
![image](https://github.com/godotengine/godot-proposals/assets/39263675/807825f2-77f1-4ed6-a32b-f97017a19c85)

**Comment by AThousandShips** [comment]
See:
* https://github.com/godotengine/godot/issues/93168

**Comment by azm-gh** [comment]
What is the timeline when it comes to namespace support? I came from this godot talk: [](https://youtu.be/SsfSZN_BEjQ?t=868)

**Comment by Calinou** [comment]
> What is the timeline when it comes to namespace support please? 

There is no ETA for implementing namespaces, as nobody has started work on them yet to my knowledge.

**Comment by brohd11** [comment]
One thing that could be nice would be an 'import' or 'using' keyword. It could act exactly as const, but give a type color to denote that it is an 'imported' const.

I'm doing the single script with all my classes as preloads in my plugin, but it gives you a white name.
Being able to differentiate these from actual constants would be great for readability.

I'm pretty happy with it's function as a namespace, this would just be QoL

import Classes = preload("")
import MyClass = Classes.MyClass

**Comment by clankill3r** [comment]
If this ever gets implemented in whatever form, then it would also be nice to be able to prevent a namespace to be able to refer to another namespace. I'm gonne explain my case why I think this is good.

In my project I have the folders `src/engine` and `src/game`.

I made the mistake by putting this line in my engine:

`var fps_controller_xyz_instance : FPS_Controller_XYZ = fps_controller_xyz_packed_scene.instantiate()`

It should have been:

`var fps_controller_xyz_instance : CharacterBody3D = fps_controller_xyz_packed_scene.instantiate()`

I won't go in to further detail but long story short, it will stop on certain lines and give errors, but not about the file that was causing the problem. I had to duplicate the project and keep removing code till I isolated it which was a huge task.

(Better out of order compilation could fix this as well...)

**Comment by Shadows-of-Fire** [comment]
I would like to enumerate what I think namespaces could look like to avoid potential future decisions being made at implementation time. I was thinking about just doing it, but there's quite a lot not specified.

# Language Semantics
## Namespaces
A **namespace** is an identifier which groups one or more classes into a named category. Namespaces provide the following benefits:
* Avoids putting all `class_name` definitions into the global namespace (the empty string).
* Provides locality for related classes.
* Allows documenting groups of classes by providing documentation for the whole namespace.

A script file may optionally define which namespace it is a member of by using the `namespace` keyword _as the first expression_ in the script. Unless explicitly declared, a class is in the global namespace and risks collision with other classes in the global namespace.

Namespaces may only use ASCII letters, numbers, and underscores in identifiers.

Namespaces may be nested by concatenating namespace names with dots (`.`).

The Godot engine reserves the `godot` namespace and any nested namespaces beneath it. User defined scripts may not add new classes or nested namespaces to it.

Namespaces should follow the following naming conventions, which make them immediately identifiable when reading code that relates to them:
* Namespaces should use `snake_case`.
* Namespaces should not begin with a number.
* When creating a new namespace, users should prefer the [reverse-domain-name convention](https://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html) used by the Java programming language for their base namespace names.

As a forward-thinking note, we should reserve a way to attach annotations or other metadata to namespaces.

## Creating Namespaced Classes
To declare a class as a member of a namespace, the following syntax must be added _as the first expression in a script file_:
```gdscript
namespace <namespace identifier>
```

Once applied, the entire script file is a member of that namespace. This has the following side effects:
* Classes in that script file (using `class_name` or ` class`) become a member of that namespace.
* Classes in that namespace are automatically imported and are available without qualifiers in that script file.
  * This means that `namespace <namespace identifier>` _implies_ `import <namespace identifier>`. Imports are defined later.

In the core, a class's Fully-Qualified Class Name (FQCN) will include the namespace (in addition to the file and the class name). For the remainder

## Accessing Namespaced Classes

Access to namespaced classes is possible through two mechanisms. The first is through a Fully-Qualified Class Name (FQCN). Any expression that currently accepts a class name (e.g., `ClassA`, `ClassA.Inner`) may now accept the FQCN to lookup a namespaced class.

A FQCN is the full namespace name, followed by a dot, followed by a class name.

The second mechanism is **import** statements. Import statements declare that in the current scope, all target(s) from that import are brought in-scope and may be accessed directly without the FQCN. Similar to how we currently access classes in the global scope.

Imports may take one of the following forms:

* `import <namespace identifier>.*` (wildcard import) - Imports all classes from a namespace into the current scope.
* `import <namespace identifier>.ClassName` (direct import) - Imports _a single class_ from a namespace into the current scope.

Conventionally, users should prefer single-class imports. The IDE should largely mask the burden of working with imports through suggestions, autocomplete, and collapsing of the imports block.

For simplicity, import statements may only be used at the top of a script file, after the `package` declaration, but before a `class_name` declaration. This may be relaxed in the future to permit method-local imports, but is not necessary for a basic implementation (and vastly simplifies name resolution rules due to it being per-file).

When using direct imports, importing two classes (from different namespaces) with the same class name is an error.

When using multiple wildcard imports, name collisions are handled by last-import-wins. Name collisions can be overridden with a following direct import. If a direct import already exists for a class name (prior to the wildcard import), the direct import takes precedence.

While only one name may be imported when dealing with conflicts, the other class may still be accessed via FQCN.

Class names that are imported take precedence over those in the global scope. In the future, we should migrate all code out of the global scope, and not permit adding new classes to the global scope.

# Editor Behavior
## Namespaces

The editor should have first-party support for namespaces. When creating a new script file, the editor should _automatically_ prompt the user to declare a namespace _and_ a class name for the script.

When the editor is suggesting a class name, the editor should support the following behavior:
* Suggesting new (not yet imported) class names (by only typing the class name, not the FQCN) and automatically importing them.
* Prioritizing existing imported class names when making suggestions.
* Suggesting continuations for FQCNs when a FQCN is detected (by entering `<root namespace>` plus a dot (`.`))

# Examples
A simple class in a namespace
```gdscript
# TestFile.gd
namespace dev.shadowsoffire.test

class_name TestFile extends RefCounted

func do_something() -> void:
    pass
```

Using classes from the same namespace
```gdscript
# TestThing.gd
namespace dev.shadowsoffire.test

class_name TestThing extends RefCounted

var file: TestFile = null

func _ready() -> void:
    self.file = TestFile.new()
```

Using classes from other namespaces
```gdscript
# OtherThing.gd
namespace dev.shadowsoffire.others

import dev.shadowsoffire.test.TestThing

class_name OtherThing extends RefCounted

@export var thing: TestThing # Legal, TestThing is imported

@export var file: TestFile # Error. TestFile is not imported, so we don't know about it.

@export var file2: dev.shadowsoffire.test.TestFile # Legal, access via FQCN.
```
___

I have more thoughts, but this is getting a bit long. I'll stop here to open to others for discussion. I think this is a good formal basis for implementing namespaces and the expected behaviors, should there be consensus on this approach.

**Comment by yookoala** [comment]
@Shadows-of-Fire Thanks for the draft! This is largely great work. 

If I may, I would suggest to use an approach that mimic Python, which many of GDScript features are inspired from. 

To mimic Python, the imports are either one below:

Direct import (You MUST use the whole path to reference the symbol):

```gdscript
import dev.shadowsoffire.test.TestThing

class_name OtherThing extends dev.shadowsoffire.test.TestThing
```

Or from-import:

```gdscript
from dev.shadowsoffire.test import TestThing

class_name OtherThing extends TestThing
```

Or from-import-as, where you can alias names:
```gdscript
from dev.shadowsoffire.test import TestThing as T

class_name OtherThing extends T
```

Or wildcard import:
```gdscript
from dev.shadowsoffire.test import *

class_name OtherThing extends TestThing,
```

P.S. Wildcard import is perceived as a bad practice in Python and should be avoided. So it would be even cleaner if we do away wildcard import in GDScript to prevent it from happening at all.

**Comment by yookoala** [comment]
Also I might add that the current proposal does not specify how the referenced code is loaded before referencing. 

If I am not mistaken, currently the dependencies of a script are loaded as per `@preload` calls specified. Unlike Python, there is no relation between a namespace path and a filesystem path. And adapting the whole finder+loader mechanics would seems like overkill.

**Comment by dog-on-moon** [comment]
Alongside declaring namespaces as `namespace <namespace identifier>`, my friend wonders if it would be plausible to adapt the current class name syntax as follows:

```gdscript
class_name Character.Rig.TorsoController

# ... as an alias for ...

namespace Character.Rig
class_name TorsoController
```

There's a couple benefits to this:
1. This is a reasonably natural extension of how class name access already works (like accessing inner classes with GlobalClass.Inner)
2. Easy to refactor -- for example, I can global find-and-replace "TorsoController" to "Character.Rig.TorsoController" (scanning whole words and match case) to simultaneously update the script's namespace and explicitly reference it everywhere via FQCN.

But it can be useful in its own right to assign a namespace to the entire script (like defining a bunch of inner classes in one script). It also makes it less explicit that you're able to access other names in `Character.Rig`. Regardless, I wonder if it's reasonable to have both?

---

I think imports are necessary too. I'm preferential to supporting wildcards since they're fast and convenient with some addons (like, if an addon was just 3 classes you'd use together commonly, it'd be nicer to just go `import that_addon.*`).

Also, one of the reasons Python's wildcard imports are bad is because of how they can import at extreme depths (like if you wildcard import file A, which wildcard imports file B, which wildcard imports file C, then everything breaks if file C tries to import anything from file A!). With the way namespaces are setup, I think this issue is avoided?

On that note, would namespaces have implied importing like this?

```gdscript
# apple.gd
namespace Food
class_name Apple

import Animals.Worm

# toast.gd
namespace Food
class_name Toast

var worm: Apple.Worm # <== valid?
```

Supporting a way to alias imports (like from-import-as) is sensible, but could be done without adding the from keyword (e.g. `import Character.Rig.TorsoController as TC`)

**Comment by dog-on-moon** [comment]
OK, important addendum (and sanity check) because I keep confusing myself:

In python, if you wildcard import from a file, you import everything from that file -- variables, functions, classes, and (unfortunately) other nested imports.

With this proposal, if you wildcard import, you *only* import the class names associated with the parent namespace:
- `import Animals.*` will import every class in the Animals namespace
- `import Animals.Worm` imports specifically the worm from the Animals namespace
- `import Animals.Worm.*` imports every class in the Animals.Worm namespace
- But unlike Python, you cannot put the *contents* of Animals.Worm into the accessible namespace.
  - You will have to go through the Worm first.

Therefore, imports are necessarily different from Python in a subtle way -- you aren't importing the *contents* of scripts, you're permitting *access* to classes. This is good, and helps keep namespaces be purely for organizing class access. (I think this means that wildcard imports are less dangerous than in Python, but I struggle to predict this.)

**Comment by Shadows-of-Fire** [comment]
> from-import:

I don't think we need multiple import syntaxes. The IDE should manage them and they should be effectively invisible. In python they're more manual, and `from x import a,b,c` allows you to condense them.

> from-import-as

I intentionally omitted any kind of aliasing to avoid overlap with https://github.com/godotengine/godot-proposals/issues/11503
It would be trivial to permit additional import-alias syntax once a mechanism for aliasing has been developed.

> P.S. Wildcard import is perceived as a bad practice in Python and should be avoided. So it would be even cleaner if we do away wildcard import in GDScript to prevent it from happening at all.

The _reasons_ why it is bad practice in Python (pollution of the global namespace) aren't applicable here, since you can only import class names (and everything else is class-bound, as it is in Java).
___
> Alongside declaring namespaces as namespace <namespace identifier>, my friend wonders if it would be plausible to adapt the current class name syntax as follows:

That sounds _possible_ but might make the grammar more complicated than it needs to be, since `class_name <FQCN>` needs to be expanded to `namespace <namespace identifier> class_name <class name>`

> On that note, would namespaces have implied importing like this?

No, I don't think that makes sense. Imports are transient, meaning they are used during name resolution local to a file and do not exist otherwise. I think what you're looking for falls under type aliases (https://github.com/godotengine/godot-proposals/issues/11503)

> With this proposal, if you wildcard import, you only import the class names associated with the parent namespace:

Correct.

**Comment by luismasuelli** [comment]
Any hint on when will this be a thing?

Either using the separate syntax:

```
namespace Foo.Bar
class_name Baz
```

Or the complete syntax:

```
class_name Foo.Bar.Baz
```

But please any kind of roadmap will be useful. This topic becomes increasingly annoying considering that we want to avoid pollution in production code when importing multiple packages.

In particular, I made a hack where I can import nested objects with `const bar = preload("bar/index.gd")` then `const Baz = preload("baz.gd")` and then have Foo as the only thing with `class_name Foo` but, still, this hack finds its limitation when you realize that you cannot use in the editor any Resource subclass declared like this:

```
# my-type.gd
extends Resource

@export var a: int
@export var b = String

# e.g. some-node-script.gd
extends Node2D

const MyType = preload("./my-type.gd")
@export var glorg: MyType
```

**Comment by obsurveyor** [comment]
Just ran into a collision with an addon today, would really like to see namespaces happen soon.

**Comment by Shadows-of-Fire** [comment]
We plan on determining the feasibility of applying namespaces to the Godot 4.x core soon. Without being able to break compatibility some things might be "weird", but given that inner classes already exist by having a class name of `Outer.Inner` it should be _feasible_ to add them.

The implementation may take a while. The language semantics are pretty easy, but getting all the information where it needs to be and updating the editor to support a namespaced world is a bit more difficult.

___

> Also I might add that the current proposal does not specify how the referenced code is loaded before referencing.

@yookoala What changes do you anticipate to the loading mechanisms? I don't anticipate any. Namespaces / imports are basically shorthand for using `Name` instead of `fully.qualified.class.Name` in a given script file. Once the imports have been lowered to FQCN's the remaining language semantics are the same.

**Comment by yookoala** [comment]
> We plan on determining the feasibility of applying namespaces to the Godot 4.x core soon. Without being able to break compatibility some things might be "weird", but given that inner classes already exist by having a class name of `Outer.Inner` it should be _feasible_ to add them.
> 
> The implementation may take a while. The language semantics are pretty easy, but getting all the information where it needs to be and updating the editor to support a namespaced world is a bit more difficult.
> 
> > Also I might add that the current proposal does not specify how the referenced code is loaded before referencing.
> 
> [@yookoala](https://github.com/yookoala) What changes do you anticipate to the loading mechanisms? I don't anticipate any. Namespaces / imports are basically shorthand for using `Name` instead of `fully.qualified.class.Name` in a given script file. Once the imports have been lowered to FQCN's the remaining language semantics are the same.

Correct me if I am wrong.

Right now, there is no way to import a GDScript file by the class name. There is no implied rules to do it. So if we want this to work:

```
import <namespace identifier>.
```

We either:
1. need rules to do implied namespace-path discovery to include the relevant file; or
2. need to explicitly have user do a `preload()` call to load relevant file; or
3. need to change Godot behaviour to index all class files in the project to correctly load the relevant file; or
4. need user to manually specify the depending class file(s) to preload globally / to component / to the script in the UI.

And the specification with an `import <namespace identifier>` statement will need to specify which of the above, or other solution to do with an import.

**Comment by Shadows-of-Fire** [comment]
> Right now, there is no way to import a GDScript file by the class name.

I'm... not entirely sure what you mean by that.

```
import <namespace identifier>.<wildcard or classname>
```
is syntax sugar for using `<class name>` in a given script file over `<fully qualified class name>`. It does not have any other semantics than that.

It is not a C-style `#include` directive, nor does it carry the "weight" of a Python `import` which does some initialization logic.

**Comment by DevSavvySerge** [comment]
> > **If this enhancement will not be used often, can it be worked around with a few lines of script?:**
> > No it can't
> 
> > especially if you want to keep having each class separately (i.e. don't use nested classes)
> 
> On the project level, it's kind of possible, since you can create a single global class to load other classes (not necessarily inner classes), see concrete example in [`WAT/namespace.gd`](https://github.com/CodeDarigan/WAT/blob/fa509c093dbbd1bb71f2f075aa091aa6bdd8667e/addons/WAT/namespace.gd).
> 
> # utils.gd
> 
> # Note: no class_name here!
> extends Node
> 
> static func say_hello():
> 	print("Hello!")
> 
> # namespace.gd
> class_name Game
> 
> const Utils = preload("res://utils.gd")
> 
> # main.gd
> extends Node2D
> 
> func _ready():
> 	Game.Utils.say_hello()
> 
> Not sure how efficient such solution would be compared to just using `class_name` at run-time, since it takes an extra "namespace resolution" with such approach.
> 
> Avoiding using too many global classes can also speed up editor performance in Godot 3.2+: [godotengine/godot#27333](https://github.com/godotengine/godot/issues/27333).
> 
> But yeah namespace issues in general are must have to resolve if we aim for proper plugins support, and we have similar issues with C++ namespaces when it comes to exposing core classes to scripting, with some solutions: [godotengine/godot#26990](https://github.com/godotengine/godot/pull/26990).

Nice workaround.

Any way to have the correct static typing working with this?

```
# namespace.gd
class_name Game

const Utils : ??? = preload("res://utils.gd") # need a typed GDSCript?
```

**Comment by derkork** [comment]
Technically this is a GDScript, so

```gdscript
const Utils:GDScript = preload("res://utils.gd")
```

**Comment by dalexeev** [comment]
> Any way to have the correct static typing working with this?

When you `preload` a type (class or enum), you don't need to specify an explicit type (`GDScript` or `Script`), as this removes the internal `is_meta_type` flag and can make it impossible to use the constant as a type hint. Type inference (`:=`) for constants also makes no difference, so the correct option in the current system is:

```gdscript
const Utils = preload("res://utils.gd")
```

It is not possible to explicitly specify the correct type here because Godot/GDScript does not support higher-order types.

**Comment by cnwangjie** [comment]
Name conflicts are very common when using addons or plugins. I can control it in my own code, but many libraries will use some common name like `Util` or some common additional implementations like `Set`. I don't really need namespace. But it's needed to have some mechanism to limit code scope like `mod` in Rust or `package` in Go.

**Comment by hayahane** [comment]
This may also help constraint what a script can do. For example, if I want to use gdscript for a level flow, so I can ban APIs like FileAccess.

**Comment by IntangibleMatter** [comment]
@hayahane While that sounds like a really good use case for modding and the like, I'm not sure if that'd be out of scope for an initial implementation or not.

Personally I like the idea, I just don't know if it's relevant to getting the core implemented.

Also how would you constrain namespaces? Would you add something in the definitions? Would it be in the project settings? Where and how would you configure that?

**Comment by Ivorforce** [comment]
I think access restrictions are out of scope for namespaces, as the concepts don't overlap at a technical level.
However, @hayahane's general idea (I'd say) has been proposed and discussed in more detail in https://github.com/godotengine/godot-proposals/issues/5010.

**Comment by Shadows-of-Fire** [comment]
Access restrictions in Java (done at the module level) are implemented in a way that depends on namespaces being present. Nonwithstanding, we must keep it out of scope here. Namespaces are a critical enough feature that they need to not be stapled to other adjacent features for the first pass.

**Comment by glenfletcher** [comment]
**using** and **using namespace** (from c++) may be better than **import** for loading names from namespaces, it better describes how this would likely work in GDScript. This would also work well with #14465, since the global functions would be migrated into namespaces. As for renaming a loaded name, we could have **using X as Y** or **using Y = X**.

**Comment by fenilli** [comment]
All use cases talked are about addons, but there is also one that is a bit of a tricky one, that is Built-in global types.

Right now if you try to do something like `class_name Logger` you will get `Class "Logger" hides a native class`, and that is a bit of a bummer specially when you want to make something that the current native class can't but the name makes sense to exist inside your codebase.

**Comment by fenilli** [comment]
I would like to add to the proposal in terms of file wide modules, meaning that we can separate a module into multiple files, and I would call this proposal more like a module system and not pure namespacing.

My proposal for this is to not use new keywords but to use annotations `@` for `@module` and also always use module fullpath to help with ambiguity, as to how this could be done in GDScript, we already have a global namespace, we could just use it to enable the modules, we already can do with inner classes, so the difference is only adding things under the same scope by name in that global namespace, and the outer is kinda like a `@abstract` that can't be instantiated but in `@module` case can't be used anywhere, it is just a path, and probably also modules can only be used with `class_name` too.

```py
# file1.gd
@module("Foo")
class_name Something


# file2.gd
@module("Foo")
class_name Else


# file3.gd
## as you can see no need for modules for all files, this is global module ( like the way it is right now Root.Entirely )
class_name Entirely

func entirely() -> void:
  Foo.Something.new() # see Foo at start, like class_name and inner class syntax, module is just adding to global namespace Foo with Something
  Foo.Else.new() # they are the same module in different files, but added to the same namespace
```

**Comment by brohd11** [comment]
So... anyone have any namespace strategies they can share? Extending from a preload is getting patched out in 4.8. A global class extending from anything that isn't using a string script path/uid throws an error since 4.6, not good for a plugin you want to distribute
```gdscript
class_name MyClass
extends Global.MyClassBase # allowed, but every class that does this prints an error when the editor starts

const MyClassBase = preload("x") # not allowed to extend

const MyClasses = preload("x")
const MyClassBase = MyClasses.MyClassBase # not allowed to extend
```
I'm running out of options here... I'll take anything that stops me from having to use paths :|

**Comment by fenilli** [comment]
> So... anyone have any namespace strategies they can share? Extending from a preload is getting patched out in 4.8. A global class extending from anything that isn't using a string script path/uid throws an error since 4.6, not good for a plugin you want to distribute

Back to C namespacing, basically just add prefixes to all classes.

```gdscript
# your plugin name (NicePlugin)

class_name NPClass
class_name NPModuleClass
class_name NPModuleSecondClass
```

**Comment by luismasuelli** [comment]
brohd11 makes total sense. I use the same strategy for namespaces. Moving it to prefixing is less practical, annoying in terms of verbosity, prone to errors in case of a refactor, and a lost opportunity in terms of ux.

Obtener Outlook para Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by neikeq** [comment]
Adding namespaces as described in this conversation is looking at the problem from the wrong angle, because the script path (or UID) already acts as a namespace for the script class. The actual problem is the way global classes were implemented.

The way global classes should work is by being identified by the script resource itself, not by the class name. This would allow adding multiple scripts with the same class name.

On the editor side, nothing changes from the user's perspective, except that the class list ("Add Node" dialog, "Create Resource" dialog, etc) now displays the script path alongside the name (either always or only when disambiguation is needed). In the scene tree and the inspector (and everywhere else) it's still displayed as if it was a built-in type (which is the selling point of global classes). Internally, the script is serialized in the scene/resource file with the script path instead of the class name.

On the GDScript language side, you would need to make it possible to disambiguate when there are two classes with the same name, e.g.: `const NPC = preload("res://addons/addon_1/NPC.gd")` (ignores globals and just uses that one).

Alternatively, you might want to stop global classes from being visible to GDScript automatically, and just make the user import them like in any other language, but that's very controversial and can't be done without breaking compat.

**Comment by jepessen** [comment]
The implementation should not be so hard.
It should be something like using
```
@namespace Core.Data
```
at the top of a class file for defining the namespace, and something like
```
@import Core.Data.*
@import Core.Data.Player
```
for importing namespaces into the file.

I should avoid more complicate things like defining different namespaces in the same file like it's possible to do in C++, but I strongly agree that the using of namespaces in a proper way really important for maintaining the code organized. The C-style with prefixes is old, make the code less readable and it's very error-prone during operations like refactoring.

I don't agree with @neikeq when he saus that the UID acts like a namespace. It's a way for gdscript to recognize a resource even if it's moved, but it's not the right tool for writing code. If I want to inherit a class, I want to inherit Core.Player, not `uid://bpuau57rkq325`, because it's totally unreadable and when I see it in the code I don't have any idea about the base class.

Namespaces are a must-to-have, and after that a plugin that allows to refactor automatically the code by giving to the script a namespace cohereng with the folder path, like C# does.

**Comment by neikeq** [comment]
> If I want to inherit a class, I want to inherit Core.Player, not `uid://bpuau57rkq325`, because it's totally unreadable and when I see it in the code I don't have any idea about the base class.

The UID is an internal thing. My code example uses paths for importing, similar to other languages. And it's only used when disambiguation is needed (assuming we kept implicit imports).

In any case, I'm not so confident with this solution any more. I've come to realize that, while this issue only mentions GDScript, we probably want to support namespaces for GDExtension as well, which are not resources. And having two different namespace systems is not a good idea.
